sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala

```diff
@@ -223,8 +223,9 @@ trait ExpressionEvalHelper extends GeneratorDrivenPropertyChecks with PlanTestBa
           }
         } else {
           val lit = InternalRow(expected, expected)
+          val dtAsNullable = expression.dataType.asNullable
```

Comment:
I didn't go through the entire thread. But my opinion is that, the data type nullable should match the real data.

BTW will this reduce test coverage? It seems the optimization for non-nullable fields is not tested if we always assume the expression is nullable.

> we use the default value for the type, and create a wrong result instead of throwing a NPE.
This is expected and I think it's a common symptom for nullable-mismatch problems. Why can't our test expose it?

Reply:
@ueshin @cloud-fan Thank you for good summary.

I think that this does not reduce test coverage.  
This ` dtAsNullable = expression.dataType.asNullable` is used only for generating `expected`. This `asNullable` does not change `dataType` of `expression`. Thus, this does not change our optimization assumption.